### DISCUSSION TITLE
524 Error from Cloudflare while loading big file for embedding.

### DISCUSSION THREAD

**Discussion by alesha-pro** [open-webui/open-webui#6648] [Issues]
# Bug Report

I am trying to upload a 340 page pdf file (book) to the knowledge base, about 14mb. I am using ollama to build embeddings. The process is going on, I can see in the logs that the chunks are vectorized, but the connection is open in the browser. That is, the request has gone and is waiting for a response until the whole file is successfully vectorized and a confirmation is sent to the frontend. Now the connection fails with 524 error, but the embedding process continues on the backend.

## Installation Method

Helm Chart installation 

## Environment

- **Open WebUI Version:** v0.3.35

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [] I have included the browser console logs.
- [ ] I have included the Docker container logs.
- [x] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:

The file is successfully uploaded to the knowledge base.

## Actual Behavior:

I'm getting error 524 from Cloudflare because a timeout has passed and the connection is still open.

## Reproduction Details

**Steps to Reproduce:**
I use DNS from Cloudflare try uploading a large file to the knowledge base.

## Additional Information

In my opinion the problem is pretty clear. You could use some web hack to notify the frontend that a file has been successfully uploaded to the vector storage, or something like that. Lots of ways.

## Screenshot
![image](https://github.com/user-attachments/assets/6a998789-a423-4527-b1d4-d5d35b5668ec)

**Comment by alesha-pro** [comment]
??

**Comment by lynxul25** [comment]
Just throwning this here, Cloudflare has a 60s limit on connections. All plans up to enterprise have this limitation which you can also see if you generate images that take longer than 60s to generate.

**Reply by alesha-pro** [reply]
Sounds bad

**Comment by ArchonMegalon** [comment]
So - why not take this approach?

https://towardsdatascience.com/how-to-fix-a-cloudflare-gateway-timeout-error-in-a-flask-vue-js-ml-application-efe2826585bb

After all, you guys are implementing both the ollama and the open webui, don't you?

**Reply by eric2788** [reply]
> So - why not take this approach?
> 
> [towardsdatascience.com/how-to-fix-a-cloudflare-gateway-timeout-error-in-a-flask-vue-js-ml-application-efe2826585bb](https://towardsdatascience.com/how-to-fix-a-cloudflare-gateway-timeout-error-in-a-flask-vue-js-ml-application-efe2826585bb)
> 
> After all, you guys are implementing both the ollama and the open webui, don't you?

yeah I wish they can implement this

**Comment by pgallagher07** [comment]
Here's a bit of a brainstorm that I had when researching this issue and reviewing the notes & links above. 

It seems like it would help to update this code in the frontend API to capture a 202 code and the uuid of the file, and start polling for a 200 code. We'd have to also make sure we gracefully handle any side effects (i.e. if the backend fails to finish processing, etc.). https://github.com/open-webui/open-webui/blob/a3bb7df61058e690a76cebb7681bd5390e77d226/src/lib/apis/files/index.ts#L16-L18

Then in this function on the backend side, where `if process` starts begins a series of steps that could cause a delay in returning a response. `if process` should call a backend process in a different thread (i.e. how the linked article provided by @ArchonMegalon uses a separate thread) and return a 202 code.  We would need a new endpoint in this file to poll that would receive the uuid as the identifier and return a 202 or 200 code; if 200, it should also return what the function currently returns now. the frontend api could poll for this and would consistently get an immediate response, preventing the 524 error from generating.
https://github.com/open-webui/open-webui/blob/a3bb7df61058e690a76cebb7681bd5390e77d226/backend/open_webui/routers/files.py#L84-L180

With @lynxul25 mentioning this happens with image generation as well, I would think that this approach may be applicable to the functions that handle that activity.

**Comment by pgallagher07** [comment]
I created a fork to start a new branch to put some thoughts behind this. I'm sure more edits are needed, and I'll need to spend more time with the toolset to get comfortable debugging this; in the meantime, I thought I'd share an in-progress / first draft attempt at fixing this, if anyone wants to suggest edits or test or otherwise; hoping this helps someone.

https://github.com/pgallagher07/open-webui/tree/pgallagher/cloudflare-524

Basically, the changes I started working on are:

1. In python (FastAPI backend side) the file is returned once uploaded, without waiting for the processing of the file to finish. 
2. The python step that handles the processing has moved to a background task.
3. I added a "processed" boolean metadata tag to the files. If it's supposed to process, this is set to false until it is finished processing.
4. The step in svelte on the frontend that handles the upload will poll for the metadata on the files, continuing in a loop until it sees that process is set to true.

**Comment by apensotti** [comment]
@pgallagher07 @lynxul25 I assume there wasn't a solution for this, as it's a timeout on Cloudflare side. Did anyone get this working???Probably should just spin up a nginx proxy instead of a Cloudflare tunnel, huh?

**Reply by tjbck** [reply]
Related: #15023

**Reply by pgallagher07** [reply]
Yes I started to work on this and life, work got in the way... but I did see that someone with more skill than myself got this working with the new `process_in_background` set to default. I see that it was completed by [tjbck](https://github.com/tjbck) in #16821 - awesome job. 
I uploaded an hour-long podcast, which gets auto-transcribed, so it takes a bit of time to finish and I find to be a good test of this. It did not timeout at the 60 second mark and was pleasantly surprised that it was able to process it completely.

I think that the solution implemented was better than the path I was going down. It used an EventStream and printed the status at regular intervals until it was complete, which I assume makes Cloudflare happy. I have much to learn on the HTTP side of things. 
<img width="967" height="502" alt="image" src="https://github.com/user-attachments/assets/f46b4915-8ec4-4584-bed4-42761d8fb19b" />
